src/goto-symex/goto_symex.cpp

```diff
@@ -665,3 +681,179 @@ bool goto_symext::constant_propagate_delete_char_at(
 
   return true;
 }
+
+bool goto_symext::constant_propagate_case_change(
+  statet &state,
+  symex_assignt &symex_assign,
+  const function_application_exprt &f_l1,
+  bool to_upper)
+{
+  const auto &f_type = to_mathematical_function_type(f_l1.function().type());
+  const auto &length_type = f_type.domain().at(0);
+  const auto &char_type = to_pointer_type(f_type.domain().at(1)).subtype();
+
+  const refined_string_exprt &s = to_string_expr(f_l1.arguments().at(2));
+  const auto s_data_opt = try_evaluate_constant_string(state, s.content());
+
+  if(!s_data_opt)
+    return false;
+
+  array_exprt string_data = s_data_opt->get();
+
+  auto &operands = string_data.operands();
+  for(auto & operand : operands)
+  {
+    int character = std::stoi(id2string(operand.get(ID_value)), nullptr, 16);
+    if (isalpha(character))
+    {
+      char buffer[33];
+      if (to_upper)
+      {
+        if(islower(character))
+        {
+          sprintf(buffer, "%x", toupper(character));
+          operand.set(ID_value, buffer);
+        }
+      }
+      else
+      {
+        if(isupper(character))
+        {
+          sprintf(buffer, "%x", tolower(character));
+          operand.set(ID_value, buffer);
+        }
+      }
+    }
+  }
+
+  const constant_exprt new_char_array_length =
+    from_integer(string_data.operands().size(), length_type);
+
+  const array_typet new_char_array_type(char_type, new_char_array_length);
+  const array_exprt new_char_array(std::move(operands), new_char_array_type);
+
+  assign_string_constant(
+    state,
+    symex_assign,
+    to_ssa_expr(f_l1.arguments().at(0)),
+    new_char_array_length,
+    to_ssa_expr(f_l1.arguments().at(1)),
+    new_char_array);
+
+  return true;
+}
+
+bool goto_symext::constant_propagate_replace(
+  statet &state,
+  symex_assignt &symex_assign,
+  const function_application_exprt &f_l1)
+{
+  const auto &f_type = to_mathematical_function_type(f_l1.function().type());
+  const auto &length_type = f_type.domain().at(0);
+  const auto &char_type = to_pointer_type(f_type.domain().at(1)).subtype();
+
+  const refined_string_exprt &s = to_string_expr(f_l1.arguments().at(2));
+  const auto s_data_opt = try_evaluate_constant_string(state, s.content());
+
+  if(!s_data_opt)
+    return false;
+
+  auto &new_data = f_l1.arguments().at(3);
+  auto &old_data = f_l1.arguments().at(4);
+
+  // If it's a symbol we're using the single character override.
+  if (new_data.id() == ID_symbol)
+  {
+    const auto &oldChar = to_symbol_expr(new_data);
+    const auto &newChar = to_symbol_expr(old_data);
+
+    const auto new_char_pointer =
+      state.propagation.find(newChar.get_identifier());
```

Comment:
Can use `try_evaluate_constant()` instead which does the lookup in the propagation map and returns an optional `constant_exprt`.